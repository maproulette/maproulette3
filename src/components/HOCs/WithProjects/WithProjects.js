import { connect } from 'react-redux'
import { denormalize } from 'normalizr'
import _compact from 'lodash/compact'
import _map from 'lodash/map'
import _get from 'lodash/get'
import { projectSchema } from '../../../services/Project/Project'
import { fetchProjectChallenges } from '../../../services/Challenge/Challenge'

const mapStateToProps = (state, ownProps) => ({
  projects: _compact(_map(_get(state, 'entities.projects', {}),
    project => project.id ?
              denormalize(project, projectSchema(), state.entities) :
              null))
})

const mapDispatchToProps = dispatch => ({
  fetchProjectChallenges: (projectId) => dispatch(fetchProjectChallenges(projectId)),
})

const WithProjects =
  WrappedComponent => connect(mapStateToProps, mapDispatchToProps)(WrappedComponent)

export default WithProjects
