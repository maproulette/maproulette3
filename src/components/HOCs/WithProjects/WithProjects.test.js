import { denormalize } from 'normalizr'
import { mapStateToProps, mapDispatchToProps } from './WithProjects'
import { fetchProjectChallenges }
       from '../../../services/Challenge/Challenge'

jest.mock('normalizr')
jest.mock('../../../services/Challenge/Challenge')

denormalize.mockImplementation((project, schema, entities) => project)

let basicState = null

beforeEach(() => {
  basicState = {
    entities: {
      projects: [
        {
          id: "123",
          name: "first",
        },
        {
          id: "456",
          name: "second",
        }
      ],
    }
  }
})

test("mapStateToProps maps projects", () => {
  const mappedProps = mapStateToProps(basicState)
  expect(mappedProps.projects).toEqual(basicState.entities.projects)

  expect(mappedProps).toMatchSnapshot()
})

test("mapDispatchToProps maps function fetchProjectChallenges", () => {
  const dispatch = jest.fn(() => Promise.resolve())
  const mappedProps = mapDispatchToProps(dispatch, {})

  mappedProps.fetchProjectChallenges("someProjectId")
  expect(dispatch).toBeCalled()
  expect(fetchProjectChallenges).toBeCalledWith("someProjectId")
})
