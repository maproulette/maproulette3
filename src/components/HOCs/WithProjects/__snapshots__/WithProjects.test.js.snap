// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mapStateToProps maps projects 1`] = `
{
  "projects": [
    {
      "id": "123",
      "name": "first",
    },
    {
      "id": "456",
      "name": "second",
    },
  ],
}
`;
